res/controllers/Roland_DJ-202-scripts.js

```diff
@@ -0,0 +1,1612 @@
+            }
+            this.send(0);
+        },
+        input: function(_channel, _control, value, _status, _group) {
+            if (!value || !DJ202.bindSamplerControls) {
+                return;
+            }
+            for (let i = 1; i <= 16; i++) {
+                const group = `[Sampler${  i  }]`;
+                script.toggleControl(group, "pfl");
+            }
+        }
+    });
+
+    // TR-S
+    ///////////////////////
+    const getActiveDeck = function() {
+        const deckvolume = new Array(0, 0, 0, 0);
+        let volumemax = -1;
+        let newdeck = -1;
+
+        // get volume from the decks and check it for use
+        for (let z = 0; z <= 3; z++) {
+            if (engine.getValue(`[Channel${  z + 1  }]`, "track_loaded") > 0) {
+                deckvolume[z] = engine.getValue(`[Channel${  z + 1  }]`, "volume");
+                if (deckvolume[z] > volumemax) {
+                    volumemax = deckvolume[z];
+                    newdeck = z;
+                }
+            }
+        }
+
+        return newdeck;
+    };
+
+    this.syncButtonPressed = function(_channel, control, value, _status, _group) {
+        if (value !== 0x7F) {
+            return;
+        }
+        const isShifted = (control === 0x55);
+        if (isShifted || this.syncDeck >= 0) {
+            this.syncDeck = -1;
+        } else {
+            const deck = getActiveDeck();
+            if (deck < 0) {
+                return;
+            }
+            const bpm = engine.getValue(`[Channel${  deck + 1  }]`, "bpm");
+
+            // Minimum BPM is 5.0 (0xEA 0x32 0x00), maximum BPM is 800.0 (0xEA 0x40
+            // 0x3e).
+            if (!(bpm >= 5 && bpm <= 800)) {
+                return;
+            }
+            const bpmValue = Math.round(bpm * 10);
+            midi.sendShortMsg(0xEA, bpmValue & 0x7f, (bpmValue >> 7) & 0x7f);
+            this.syncDeck = deck;
+        }
+    };
+
+    this.startStopButtonPressed = function(_channel, _control, _value, status,
+        _group) {
+        if (status === 0xFA) {
+            this.playbackCounter = 1;
+            this.playbackTimer = engine.beginTimer(500, function() {
+                midi.sendShortMsg(0xBA, 0x02, this.playbackCounter);
+                this.playbackCounter = (this.playbackCounter % 4) + 1;
+            });
+        } else if (status === 0xFC) {
+            if (this.playbackTimer) {
+                engine.stopTimer(this.playbackTimer);
+            }
+        }
+    };
+
+    this.customSamplePlayback = function(_channel, _control, value, _status, group) {
+        if (value) {
+            engine.setValue(group, "cue_gotoandplay", 1);
+        }
+    };
+};
+
+DJ202.Sampler.prototype =
+  Object.create(components.ComponentContainer.prototype);
+
+////////////////////////
+// Custom components. //
+////////////////////////
+
+DJ202.FlashingButton = function() {
+    components.Button.call(this);
+    this.flashFreq = 50;
```

Comment:
50 Hz seems a bit much - Please add a comment, what the number 50 means here.

Reply:
It's 50ms for flashing buttons, added comment